scripts/cluster-upgrade.sh

```diff
@@ -0,0 +1,161 @@
+#!/bin/bash
+
+OCP_VERSION_FROM=$1
+OCP_VERSION_TO=$2
+
+if [ -z $OCP_VERSION_FROM ];
+then
+    echo "This script can be used to get the status of all managed clusters or to upgrade clusters."
+    echo "usage: $0 status"
+    echo "usage: $0 <from> <to>"
+    echo "  example: $0 4.1.0 4.1.2"
+    exit 1
+fi
+
+if [ "$OCP_VERSION_FROM" == "status" ];
+then
+    # absense of versions implies status check
+    unset OCP_VERSION_FROM
+    unset OCP_VERSION_TO
+fi
+
+UPGRADE_STARTED=""
+UPGRADE_PROGRESSING=""
+UPGRADE_NOT_POSSIBLE=""
+UPGRADE_DONE=""
+
+NOW_EPOCH=$(date +"%s")
+
+if [ ! -z $OCP_VERSION_FROM ];
+then
+    # Verify we can actually upgrade (is this in the graph)
+    CHANNEL_NAME=$(echo "${OCP_VERSION_TO}" | sed 's/\([^.]*\.[^.]*\)\..*/stable-\1/g')
+    GRAPH=$(curl -s -H "Accept: application/json" https://api.openshift.com/api/upgrades_info/v1/graph?channel="${CHANNEL_NAME}")
+    GRAPH_INDEX_FROM=$(echo "${GRAPH}" | jq -r "[ .nodes[] | .version == \"$OCP_VERSION_FROM\" ] | index(true)")
+    GRAPH_INDEX_TO=$(echo "${GRAPH}" | jq -r "[ .nodes[] | .version == \"$OCP_VERSION_TO\" ] | index(true)")
+
+    GRAPH_EDGE=$(echo "${GRAPH}" | jq -r ".edges[] | select(.[0] == $GRAPH_INDEX_FROM) | .[1] == $GRAPH_INDEX_TO" | grep true)
+
+    if [ -z $GRAPH_INDEX_FROM ];
+    then
+        echo "Cannot upgrade from $OCP_VERSION_FROM, it is not available in upgrade graph."
+        exit 2
+    fi
+    if [ -z $GRAPH_INDEX_TO ];
+    then
+        echo "Cannot upgrade to $OCP_VERSION_TO, it is not available in upgrade graph."
+        exit 2
+    fi
+    if [ "$GRAPH_EDGE" != "true" ];
+    then
+        echo "Cannot upgrade from $OCP_VERSION_FROM to $OCP_VERSION_TO, no path exists in upgrade graph."
+        exit 2
+    fi
+
+    # Verify the target version exists as a ClusterImageSet (CIS) on the cluster
+    # (if not, the customer could never install to that target version)
+    CIS_VERSION=$(oc get clusterimageset --all-namespaces -o json | jq -r ".items[].spec.releaseImage | split(\":\")[1] | select(. == \"$OCP_VERSION_TO\")")
+    if [ "$OCP_VERSION_TO" != "$CIS_VERSION" ];
+    then
+        echo "Cannot upgrade to $OCP_VERSION_TO, it is not available as a ClusterImageSet in the cluster."
+        exit 3
+    fi
+
+    # Verify we have ClusterDeployment CRs to work with
+    if [ `oc get crd clusterdeployments.hive.openshift.io --no-headers 2>/dev/null | wc -l` == "0" ];
+    then
+        echo "ERROR: Current cluster does not have Hive installed.  Verify where you are logged in."
+        exit 1
+    fi
+fi
+
+TMP_DIR=`mktemp -d`
+OCP_PATCH="{\"spec\":{\"desiredUpdate\": {\"force\": false, \"image\": \"\", \"version\": \"$OCP_VERSION_TO\"}}}"
+if [[ $? -ne 0 ]];
+then
+    echo "ERROR: Couldn't create a temporary directory. Aborting.
+    exit 1
+fi
+# kick off the upgrades (or status check)
+for CD_NAMESPACE in `oc get clusterdeployment --all-namespaces | awk '{print $1}' | sort | uniq`;
```

Comment:
Might want to look into using `-o jsonpath` instead of piping through awk/sort/uniq? (also, `sort -u` == `sort | uniq`)

Reply:
Yea, optimizations would be nice.  I use `uniq -c` sometimes so it's habit to do the `sort | uniq` instead of `sort -u`.  The formatting bits that oc gives are nice but don't come naturally to me.. tend to toss them at other things like awk, jq, yq..

Reply:
I totally get you here. I still struggle with `-o jsonpath`! I'm not aiming for minimal forking :) Just offering some suggestions that are "native" and ones from which we could all learn. This change is up to you to dive into, or not; I'll mark resolved.